File: src/lib/commands/workspace.ts
import { apiGet, apiPost } from '../api/client';

export type GlobalConfig = {
  window: { width: number; height: number; x: number; y: number };
  defaultShell: string;
  repos: string[];
  lastActiveRepo: string | null;
};

export type RepoConfig = {
  path: string;
  lastWorktree: string | null;
  sidebarCollapsed: boolean;
};

export type Preset = {
  name: string;
  command: string;
  tabType: 'shell' | 'claude';
  env: Record<string, string>;
};

export async function getAppConfig(): Promise<GlobalConfig> {
  try {
    return await apiGet<GlobalConfig>('/config/app');
  } catch (e) {
    console.error(`getAppConfig failed: ${e}`);
    throw e;
  }
}

export async function saveAppConfig(config: GlobalConfig): Promise<void> {
  try {
    await apiPost('/config/app', config);
  } catch (e) {
    console.error(`saveAppConfig failed: ${e}`);
    throw e;
  }
}

export async function getRepoConfig(repoPath: string): Promise<RepoConfig> {
  try {
    return await apiGet<RepoConfig>(`/config/repo?repoPath=${encodeURIComponent(repoPath)}`);
  } catch (e) {
    console.error(`getRepoConfig failed: ${e}`);
    throw e;
  }
}

export async function saveRepoConfig(repoPath: string, config: RepoConfig): Promise<void> {
  try {
    await apiPost(`/config/repo?repoPath=${encodeURIComponent(repoPath)}`, config);
  } catch (e) {
    console.error(`saveRepoConfig failed: ${e}`);
    throw e;
  }
}

export async function getPresets(repoPath: string): Promise<Preset[]> {
  try {
    return await apiGet<Preset[]>(`/config/presets?repoPath=${encodeURIComponent(repoPath)}`);
  } catch (e) {
    console.error(`getPresets failed: ${e}`);
    throw e;
  }
}

export async function savePresets(repoPath: string, presets: Preset[]): Promise<void> {
  try {
    await apiPost(`/config/presets?repoPath=${encodeURIComponent(repoPath)}`, presets);
  } catch (e) {
    console.error(`savePresets failed: ${e}`);
    throw e;
  }
}

File: src-tauri/src/commands/workspace.rs
use crate::error::AppError;
use crate::services::config::{self, GlobalConfig, RepoConfig};
use log::{debug, info};

#[tauri::command]
pub fn get_app_config() -> Result<GlobalConfig, AppError> {
    config::load_global_config()
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn save_app_config(config: GlobalConfig) -> Result<(), AppError> {
    config::save_global_config(&config)
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn get_repo_config(repo_path: String) -> Result<RepoConfig, AppError> {
    config::load_repo_config(&repo_path)
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn save_repo_config_cmd(repo_path: String, config: RepoConfig) -> Result<(), AppError> {
    config::save_repo_config(&repo_path, &config)
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn get_presets(repo_path: String) -> Result<Vec<config::Preset>, AppError> {
    debug!("Loading presets for repo: {repo_path:?}");
    let config = config::load_repo_config(&repo_path)?;
    Ok(config.presets)
}

#[tauri::command]
#[allow(clippy::needless_pass_by_value)]
pub fn save_presets(repo_path: String, presets: Vec<config::Preset>) -> Result<(), AppError> {
    info!("Saving {} presets for repo: {repo_path:?}", presets.len());
    let mut config = config::load_repo_config(&repo_path)?;
    config.presets = presets;
    config::save_repo_config(&repo_path, &config)
}
